games/TicTacToe/src/main.sw

```diff
@@ -0,0 +1,97 @@
+contract;
+
+dep interface;
+dep events;
+dep errors;
+dep utils;
+
+use interface::Game;
+use events::{GameDrawnEvent, GameWonEvent, NewGameEvent};
+use errors::Errors;
+use utils::{draw, win_check};
+use core::ops::Eq;
+use std::{auth::msg_sender, logging::log};
+
+impl<T> Eq for Option<T> {
+    fn eq(self, other: Self) -> bool {
+        match (self, other) {
+            (Option::None, Option::None) => true,
+            (Option::Some(T), Option::Some(T)) => true,
+            _ => false,
+        }
+    }
+}
+
+storage {
+    // Stores the players
+    player_one: Option<Identity> = Option::None,
+    player_two: Option<Identity> = Option::None,
+    // Stores each player's moves,
+    board: StorageMap<u64, Option<Identity>> = StorageMap {},
+    // Stores move counter
+    player_turn: Option<Identity> = Option::None,
+    move_counter: u64 = 0,
+}
+
+impl Game for Contract {
+    #[storage(write)]
+    fn new_game(player_one: Identity, player_two: Identity) {
```

Comment:
Is there anything to stop one of the players (or even a 3rd party!) ruining the game by calling `new_game` half way through and resetting everything ?

Reply:
Under the `move` function, there is a condition that checks if the right player is playing, so that's reducing the risks of compromising a game, but yeah good call, I'll add a condition to make sure no external party can reset the game in the middle of it.

Reply:
The assumption was that this will run locally for now and that no other party will reset it.

The problem with adding this check is that now you need to decide when a new game can be started.

- Is it fine to start a new game only once the previous game is over?
- Is it fine to start a new game once either of the players decides that it is over?

The first point is probably okay.

`new_game()` can be called either when it has some default pre-init value or another value which indicates that the game has ended. This way any address can call the function if-and-only-if (_iff_) there is no active game atm.

This leads to the next problem of general race conditions for who gets to play. First caller gets to decide the players and thus additional functionality now needs to be introduced to either juggle multiple games at once internally or we need access control per contract.

There is also a secondary problem of a player never making a move and thus the contract is locked since `new_game()` cannot be called because the game has not reached a conclusion. Now we need additional functionality through privileges, time spent on a game etc. before a new game can be started so that the contract is not locked.

Reply:
I think this and my comment below could be solved by explicitly storing the game's state. Then it's easy to see if a game is complete or not. 

If the current design is tailored to a "friendly" game, then the risk of a player abandoning the game is lower than a malicious 3rd party resetting it. We can add a chess clock mechanism in a later PR.

Happy to defer resolution of these attack vectors to a later PR, but we should document them somewhere. (Either TODOs or preferably issues)

Reply:
What is the purpose of another storage variable?

We could use another variable; but, we can also iterate over the positions and figure out if the game has ended. With the gas estimation provided by the SDK we can check which implementation is cheaper. There's also code quality to consider. It's significantly cleaner to access a variable than to have a method to iterate and figure out a state.